bot/cogs/utils.py

```diff
@@ -173,6 +196,64 @@ def check(m: Message) -> bool:
             f"as I detected unauthorised use by {msg.author} (ID: {msg.author.id})."
         )
 
+    @command()
+    async def zen(self, ctx: Context, *, search_value: Union[int, str, None] = None) -> None:
+        """
+        Show the Zen of Python.
+
+        Without any arguments, the full Zen will be produced.
+        If an integer is provided, the line with that index will be produced.
+        If a string is provided, the line which matches best will be produced.
+        """
+        embed = Embed(
+            colour=Colour.blurple(),
+            title="The Zen of Python",
+            description=ZEN_OF_PYTHON
+        )
+
+        if search_value is None:
+            embed.title += ", by Tim Peters"
+            await ctx.send(embed=embed)
+            return
+
+        zen_lines = ZEN_OF_PYTHON.splitlines()
+
+        # handle if it's an index int
+        if isinstance(search_value, int):
+            upper_bound = len(zen_lines) - 1
+            lower_bound = -1 * upper_bound
+            if not (lower_bound <= search_value <= upper_bound):
+                raise BadArgument(f"Please provide an index between {lower_bound} and {upper_bound}.")
+
+            embed.title += f" (line {search_value % len(zen_lines)}):"
+            embed.description = zen_lines[search_value]
+            await ctx.send(embed=embed)
+            return
+
+        # handle if it's a search string
+        matcher = difflib.SequenceMatcher(None, search_value.lower())
+
+        best_match = ""
+        match_index = 0
+        best_ratio = 0
+
+        for index, line in enumerate(zen_lines):
+            matcher.set_seq2(line.lower())
+
+            # the match ratio needs to be adjusted because, naturally,
+            # longer lines will have worse ratios than shorter lines when
+            # fuzzy searching for keywords. this seems to work okay.
+            adjusted_ratio = (len(line) - 5) ** 0.5 * matcher.ratio()
+
+            if adjusted_ratio > best_ratio:
+                best_ratio = adjusted_ratio
+                best_match = line
+                match_index = index
```

Comment:
Shouldn't we use a threshold below which we say "Sorry, I couldn't find anything matching you query."? The current mechanism *always* returns something, even if the string you specify has nothing to do with any of the lines in the zen of Python. 

It does lead to interesting responses.
![2020-03-05_14-17](https://user-images.githubusercontent.com/33516116/75985195-27b9c400-5eec-11ea-8e76-8d173bddf3ed.png)

Reply:
I actually agree with you, and I pondered over it for a while. A particularly nice one is:
![image](https://user-images.githubusercontent.com/29337040/75985269-be5fa280-5f37-11ea-8388-1ac31340db3f.png)

In the end, it's become obvious that we need to implement a decent difflib based partial matching solution that's used as a utility across the entire project(s), and honestly I don't find it absolutely essential for it to be implemented in this PR to get it merged and dealt with for later enhancement.

Reply:
That's fair. We're now using fuzzy matching in a lot of different places.